File: server/backend/src/backend/api/models/season.py
from datetime import datetime

from backend.api.models.base import RequestBase, ResponseBase


# ---------- Strict input (create/update) ----------
class _SeasonAPIBaseIn(RequestBase):
    """Fields for requests: validate here if needed."""

    name: str


class SeasonAPIRequestDetached(_SeasonAPIBaseIn):
    """Request body without external relations, allowing for dynamic insertion"""

    slug: str


class SeasonAPICreate(_SeasonAPIBaseIn):
    """Request body for creating a season."""

    show_id: int
    index: int
    slug: str


class SeasonAPIUpdate(_SeasonAPIBaseIn):
    """Request body for updating a season."""

    index: int


# ---------- Lenient output (read) ----------
class _SeasonAPIBaseOut(ResponseBase):
    """Fields for responses: no validators, no constraints."""
    id: int
    show_id: int
    index: int
    name: str
    slug: str


class SeasonAPIRead(_SeasonAPIBaseOut):
    """Response body for a season."""
    created_at: datetime
    updated_at: datetime

File: ui/src/types/schemas/season.ts
import {z} from "zod";

import {ApiDateTimeSchema} from "./datetime";


// ---------- Strict request (create/update) ----------
const SeasonBaseSchema = z.object({
    name: z.string().min(1, "Name is required")
})

export const SeasonCreateSchema = SeasonBaseSchema.extend({
    showId: z.int().min(1, "showId missing"),
    index: z.int(),
    slug: z.string().min(1, "slug missing"),
})
export type SeasonCreateIn = z.input<typeof SeasonCreateSchema>;
export type SeasonCreateOut = z.output<typeof SeasonCreateSchema>;


export const SeasonDetachedSchema = SeasonCreateSchema.omit({
    showId: true,
    index: true,
})
export type SeasonDetachedIn = z.input<typeof SeasonDetachedSchema>;
export type SeasonDetachedOut = z.output<typeof SeasonDetachedSchema>;


export const SeasonUpdateSchema = SeasonBaseSchema.extend({
    index: z.int(),
})
export type SeasonUpdateIn = z.input<typeof SeasonUpdateSchema>;
export type SeasonUpdateOut = z.output<typeof SeasonUpdateSchema>;


// ------------ Lenient response (read) ------------
export const SeasonReadSchema = z.looseObject({
    id: z.int(),
    showId: z.int(),
    index: z.int(),
    name: z.string(),
    slug: z.string(),
    createdAt: ApiDateTimeSchema,
    updatedAt: ApiDateTimeSchema,
})
export type SeasonRead = z.infer<typeof SeasonReadSchema>;
